### PR TITLE
Fixes #29567 - Add nulldb for build environments

### PR DESCRIPTION
**Pull request by ehelms**

### CODE REVIEW

**Comment by ehelms** [comment]
Packaging changes: 
 - https://github.com/theforeman/foreman-packaging/pull/5059
 - https://github.com/theforeman/foreman-packaging/pull/5058

CI:
 - https://github.com/theforeman/foreman-infra/pull/1338

To do:

 - [x] Test packaging with this PR
 - [x] Test nulldb against Rails 6

File: bundler.d/nulldb.rb [hunk 1]
```diff
@@ -0,0 +1 @@
+gem 'activerecord-nulldb-adapter'
```
Comment (tbrisker) [inline bundler.d/nulldb.rb:1]: any reason to not use a group like we do for other `bundler.d` files?

Comment (ehelms) [reply bundler.d/nulldb.rb:1]: Seems activerecord loading does not like this:

```
vagrant@centos7-katello-devel-stable foreman (fixes-29567)$ bundle exec rake apipie:cache RAILS_ENV=production cache_part=resources DATABASE_URL=nulldb://nohost --trace
/home/vagrant/foreman/lib/foreman.rb:8: warning: already initialized constant Foreman::UUID_REGEXP
/home/vagrant/foreman/lib/foreman.rb:8: warning: previous definition of UUID_REGEXP was here
** Invoke apipie:cache (first_time)
** Invoke environment (first_time)
** Execute environment
Apipie cache enabled but not present yet. Run apipie:cache rake task to speed up API calls.
** Execute apipie:cache
2020-04-19 13:58:50 +0000 | Started
rake aborted!
NullDB not configured. Require a framework, ex 'nulldb/rails'
/home/vagrant/.gem/ruby/gems/activerecord-nulldb-adapter-0.4.0/lib/nulldb/core.rb:18:in `configuration'
/home/vagrant/.gem/ruby/gems/activerecord-nulldb-adapter-0.4.0/lib/active_record/connection_adapters/nulldb_adapter/core.rb:146:in `columns'
/home/vagrant/.gem/ruby/gems/activerecord-5.2.1/lib/active_record/connection_adapters/schema_cache.rb:69:in `columns'
/home/vagrant/.gem/ruby/gems/activerecord-5.2.1/lib/active_record/connection_adapters/schema_cache.rb:75:in `columns_hash'
/home/vagrant/.gem/ruby/gems/activerecord-5.2.1/lib/active_record/model_schema.rb:466:in `load_schema!'
/home/vagrant/.gem/ruby/gems/activerecord-5.2.1/lib/active_record/attributes.rb:234:in `load_schema!'
/home/vagrant/.gem/ruby/gems/activerecord-5.2.1/lib/active_record/attribute_decorators.rb:51:in `load_schema!'
/home/vagrant/.gem/ruby/gems/activerecord-5.2.1/lib/active_record/model_schema.rb:459:in `block in load_schema'
/opt/rh/rh-ruby25/root/usr/share/ruby/monitor.rb:226:in `mon_synchronize'
/home/vagrant/.gem/ruby/gems/activerecord-5.2.1/lib/active_record/model_schema.rb:456:in `load_schema'
/home/vagrant/.gem/ruby/gems/activerecord-5.2.1/lib/active_record/model_schema.rb:346:in `attribute_types'
/home/vagrant/.gem/ruby/gems/activerecord-5.2.1/lib/active_record/attribute_methods.rb:230:in `has_attribute?'
/home/vagrant/.gem/ruby/gems/activerecord-5.2.1/lib/active_record/relation/calculations.rb:196:in `block in pluck'
/home/vagrant/.gem/ruby/gems/activerecord-5.2.1/lib/active_record/relation/calculations.rb:195:in `map'
/home/vagrant/.gem/ruby/gems/activerecord-5.2.1/lib/active_record/relation/calculations.rb:195:in `pluck'
/home/vagrant/foreman/app/models/permission.rb:14:in `resources'
/home/vagrant/foreman/app/controllers/api/v2/template_inputs_controller.rb:41:in `block (2 levels) in <class:TemplateInputsController>'
/home/vagrant/.gem/ruby/gems/apipie-rails-0.5.17/lib/apipie/validator.rb:322:in `instance_exec'
/home/vagrant/.gem/ruby/gems/apipie-rails-0.5.17/lib/apipie/validator.rb:322:in `initialize'
/home/vagrant/.gem/ruby/gems/apipie-rails-0.5.17/lib/apipie/validator.rb:315:in `new'
/home/vagrant/.gem/ruby/gems/apipie-rails-0.5.17/lib/apipie/validator.rb:315:in `build'
/home/vagrant/.gem/ruby/gems/apipie-rails-0.5.17/lib/apipie/validator.rb:34:in `block in find'
/home/vagrant/.gem/ruby/gems/apipie-rails-0.5.17/lib/apipie/validator.rb:33:in `each'
/home/vagrant/.gem/ruby/gems/apipie-rails-0.5.17/lib/apipie/validator.rb:33:in `find'
/home/vagrant/.gem/ruby/gems/apipie-rails-0.5.17/lib/apipie/param_description.rb:93:in `initialize'
/home/vagrant/.gem/ruby/gems/apipie-rails-0.5.17/lib/apipie/param_description.rb:21:in `new'
/home/vagrant/.gem/ruby/gems/apipie-rails-0.5.17/lib/apipie/param_description.rb:21:in `from_dsl_data'
/home/vagrant/.gem/ruby/gems/apipie-rails-0.5.17/lib/apipie/method_description.rb:55:in `block in initialize'

```

**Comment by ehelms** [comment]
I think the right progression for this should be:

 * Introduce nulldb
 * Update builds and CI to use it
 * Drop SQLite

@mmoll I assume part of debian build needs updating for this change too?

**Comment by mmoll** [comment]
@ehelms Exactly. Let's switch the test builds first and after that I'll have a look into DEBs (probably the test changes just need to get copied over).

**Comment by ohadlevy** [comment]
just curious, I did try it using the docker built at
https://github.com/theforeman/foreman/compare/develop...ohadlevy:docker-nulldb
and hit the following error: rake aborted! NotImplementedError:
change_column is not implemented... am I missing something?

**Comment by ohadlevy** [comment]
full build logs at
```
rake aborted! NotImplementedError: change_column is not implemented
4/19/2020, 5:35:35 PM /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/connection_adapters/abstract/schema_statements.rb:624:in `change_column' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/migration.rb:890:in `block in method_missing' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/migration.rb:858:in `block in say_with_time' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/migration.rb:858:in `say_with_time' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/migration.rb:879:in `method_missing' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/migration/compatibility.rb:96:in `change_column' /home/foreman/db/migrate/20090916053824_change_host_build_default_to_false.rb:3:in `up' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/migration.rb:831:in `exec_migration' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/migration.rb:812:in `block (2 levels) in migrate' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/migration.rb:811:in `block in migrate' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:473:in `with_connection' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/migration.rb:810:in `migrate' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/migration.rb:1001:in `migrate' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/migration.rb:1311:in `block in execute_migration_in_transaction' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/migration.rb:1364:in `ddl_transaction' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/migration.rb:1310:in `execute_migration_in_transaction' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/migration.rb:1282:in `block in migrate_without_lock' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/migration.rb:1281:in `each' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/migration.rb:1281:in `migrate_without_lock' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/migration.rb:1231:in `migrate' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/migration.rb:1061:in `up' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/migration.rb:1036:in `migrate' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/tasks/database_tasks.rb:238:in `migrate' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/railties/databases.rake:86:in `block (3 levels) in <top (required)>' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/railties/databases.rake:84:in `each' /home/foreman/vendor/ruby/2.6.0/gems/activerecord-6.0.2.1/lib/active_record/railties/databases.rake:84:in `block (2 levels) in <top (required)>' /home/foreman/vendor/ruby/2.6.0/gems/rake-13.0.1/exe/rake:27:in `<top (required)>' /usr/share/gems/gems/bundler-1.17.2/lib/bundler/cli/exec.rb:74:in `load' /usr/share/gems/gems/bundler-1.17.2/lib/bundler/cli/exec.rb:74:in `kernel_load' /usr/share/gems/gems/bundler-1.17.2/lib/bundler/cli/exec.rb:28:in `run' /usr/share/gems/gems/bundler-1.17.2/lib/bundler/cli.rb:463:in `exec' /usr/share/gems/gems/bundler-1.17.2/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run' /usr/share/gems/gems/bundler-1.17.2/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command' /usr/share/gems/gems/bundler-1.17.2/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch' /usr/share/gems/gems/bundler-1.17.2/lib/bundler/cli.rb:27:in `dispatch' /usr/share/gems/gems/bundler-1.17.2/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start' /usr/share/gems/gems/bundler-1.17.2/lib/bundler/cli.rb:18:in `start' /usr/share/gems/gems/bundler-1.17.2/exe/bundle:30:in `block in <top (required)>' /usr/share/gems/gems/bundler-1.17.2/lib/bundler/friendly_errors.rb:124:in `with_friendly_errors' /usr/share/gems/gems/bundler-1.17.2/exe/bundle:22:in `<top (required)>' /usr/bin/bundle:23:in `load' /usr/bin/bundle:23:in `<main>' Tasks: TOP => db:migrate
4/19/2020, 5:35:35 PM (See full trace by running task with --trace)
4/19/2020, 5:35:35 PM == 20090714132448 CreateHosts: migrating ====================================== -- create_table(:hosts, {:id=>:integer}) -> 0.0001s -- add_index(:hosts, :source_file_id) -> 0.0001s -- add_index(:hosts, :name) -> 0.0000s -- create_table(:fact_names, {:id=>:integer}) -> 0.0000s -- add_index(:fact_names, :name) -> 0.0000s -- create_table(:fact_values, {:id=>:integer}) -> 0.0000s -- add_index(:fact_values, :fact_name_id) -> 0.0000s -- add_index(:fact_values, :host_id) -> 0.0000s -- add_column(:hosts, :mac, :string, {:limit=>17, :default=>""}) -> 0.0002s -- add_column(:hosts, :sp_mac, :string, {:limit=>17, :default=>""}) -> 0.0000s -- add_column(:hosts, :sp_ip, :string, {:limit=>15, :default=>""}) -> 0.0000s -- add_column(:hosts, :sp_name, :string, {:limit=>255, :default=>""}) -> 0.0000s -- add_column(:hosts, :root_pass, :string, {:limit=>64}) -> 0.0000s -- add_column(:hosts, :serial, :string, {:limit=>12}) -> 0.0000s -- add_column(:hosts, :puppetmaster, :string, {:limit=>255}) -> 0.0000s -- add_column(:hosts, :puppet_status, :integer, {:null=>false, :default=>0}) -> 0.0001s -- add_column(:hosts, :domain_id, :integer, {}) -> 0.0000s -- add_column(:hosts, :architecture_id, :integer, {}) -> 0.0000s -- add_column(:hosts, :operatingsystem_id, :integer, {}) -> 0.0000s -- add_column(:hosts, :environment_id, :integer, {}) -> 0.0000s -- add_column(:hosts, :subnet_id, :integer, {}) -> 0.0000s -- add_column(:hosts, :sp_subnet_id, :integer, {}) -> 0.0000s -- add_column(:hosts, :ptable_id, :integer, {}) -> 0.0000s -- add_column(:hosts, :medium_id, :integer, {}) -> 0.0000s -- add_column(:hosts, :build, :boolean, {:default=>true}) -> 0.0001s -- add_column(:hosts, :comment, :text, {}) -> 0.0000s -- add_column(:hosts, :disk, :text, {}) -> 0.0000s -- add_column(:hosts, :installed_at, :datetime, {}) -> 0.0000s == 20090714132448 CreateHosts: migrated (0.0017s) ============================= == 20090714132449 AddAuditsTable: migrating =================================== -- create_table(:audits, {:force=>true, :id=>:integer}) -> 0.0001s -- add_index(:audits, [:auditable_id, :auditable_type], {:name=>"auditable_index"}) -> 0.0001s -- add_index(:audits, [:user_id, :user_type], {:name=>"user_index"}) -> 0.0000s -- add_index(:audits, :created_at) -> 0.0000s -- add_index(:audits, [:auditable_parent_id, :auditable_parent_type], {:name=>"auditable_parent_index"}) -> 0.0000s == 20090714132449 AddAuditsTable: migrated (0.0004s) ========================== == 20090715143858 CreateArchitectures: migrating ============================== -- create_table(:architectures, {:id=>:integer}) -> 0.0001s -- create_table(:architectures_operatingsystems, {:id=>false}) -> 0.0001s == 20090715143858 CreateArchitectures: migrated (0.0002s) ===================== == 20090717025820 CreateMedia: migrating ====================================== -- create_table(:media, {:id=>:integer}) -> 0.0001s == 20090717025820 CreateMedia: migrated (0.0001s) ============================= == 20090718060746 CreateDomains: migrating ==================================== -- create_table(:domains, {:id=>:integer}) -> 0.0001s == 20090718060746 CreateDomains: migrated (0.0001s) =========================== == 20090718064254 CreateSubnets: migrating ==================================== -- create_table(:subnets, {:id=>:integer}) -> 0.0001s == 20090718064254 CreateSubnets: migrated (0.0002s) =========================== == 20090720134126 CreateOperatingsystems: migrating =========================== -- create_table(:operatingsystems, {:id=>:integer}) -> 0.0001s == 20090720134126 CreateOperatingsystems: migrated (0.0001s) ================== == 20090722140138 CreateModels: migrating ===================================== -- create_table(:models, {:id=>:integer}) -> 0.0001s -- add_column(:hosts, :model_id, :integer, {}) -> 0.0001s == 20090722140138 CreateModels: migrated (0.0002s) ============================ == 20090722141107 CreateEnvironments: migrating =============================== -- create_table(:environments, {:id=>:integer}) -> 0.0001s -- create_table(:environments_puppetclasses, {:id=>false}) -> 0.0000s == 20090722141107 CreateEnvironments: migrated (0.0002s) ====================== == 20090729132209 CreateReports: migrating ==================================== -- create_table(:reports, {:id=>:integer}) -> 0.0001s == 20090729132209 CreateReports: migrated (0.0001s) =========================== == 20090730152224 CreatePtables: migrating ==================================== -- create_table(:ptables, {:id=>:integer}) -> 0.0001s -- create_table(:operatingsystems_ptables, {:id=>false}) -> 0.0000s == 20090730152224 CreatePtables: migrated (0.0002s) =========================== == 20090802062223 CreatePuppetclasses: migrating ============================== -- create_table(:puppetclasses, {:id=>:integer}) -> 0.0001s -- create_table(:hosts_puppetclasses, {:id=>false}) -> 0.0000s -- create_table(:operatingsystems_puppetclasses, {:id=>false}) -> 0.0000s == 20090802062223 CreatePuppetclasses: migrated (0.0003s) ===================== == 20090804130144 CreateParameters: migrating ================================= -- create_table(:parameters, {:id=>:integer}) -> 0.0001s == 20090804130144 CreateParameters: migrated (0.0001s) ======================== == 20090820130541 CreateAuthSources: migrating ================================ -- create_table(:auth_sources, {:id=>:integer}) -> 0.0001s == 20090820130541 CreateAuthSources: migrated (0.0002s) ======================= == 20090905150131 CreateHostgroups: migrating ================================= -- create_table(:hostgroups, {:id=>:integer}) -> 0.0001s -- create_table(:hostgroups_puppetclasses, {:id=>false}) -> 0.0000s -- add_column(:hosts, :hostgroup_id, :integer, {}) -> 0.0001s -- add_column(:parameters, :hostgroup_id, :integer, {}) -> 0.0000s == 20090905150131 CreateHostgroups: migrated (0.0003s) ======================== == 20090905155444 AddTypeToParameter: migrating =============================== -- add_column(:parameters, :type, :string, {:limit=>255}) -> 0.0001s == 20090905155444 AddTypeToParameter: migrated (0.0001s) ====================== == 20090907045751 AddDomainToParameter: migrating ============================= -- add_column(:parameters, :domain_id, :integer, {}) -> 0.0001s == 20090907045751 AddDomainToParameter: migrated (0.0001s) ==================== == 20090915030726 ChangeReportFieldTypeToText: migrating ====================== == 20090915030726 ChangeReportFieldTypeToText: migrated (0.0000s) ============= == 20090916053824 ChangeHostBuildDefaultToFalse: migrating ==================== -- change_column(:hosts, :build, :boolean, {:default=>false}) 
```

**Comment by ehelms** [comment]
You'll need to copy schema.rb.nulldb to schema.rb during the build

**Comment by ehelms** [comment]
The RPM PRs are up to date for the current design.

File: bundler.d/nulldb.rb (hunk 1 above)
Comment (tbrisker) [reply bundler.d/nulldb.rb:1]: Would this mean users need to install the nulldb package for foreman to start in production?

**Comment by ohadlevy** [comment]
On Mon, Apr 20, 2020 at 10:23 AM Tomer Brisker <notifications@github.com>
wrote:

> *@tbrisker* commented on this pull request.
> ------------------------------
>
> In bundler.d/nulldb.rb
> <https://github.com/theforeman/foreman/pull/7594#discussion_r411151714>:
>
> > @@ -0,0 +1 @@
> +gem 'activerecord-nulldb-adapter'
>
> Would this mean users need to install the nulldb package for foreman to
> start in production?
>
I've delete it  on my container creation branch
https://github.com/theforeman/foreman/compare/develop...ohadlevy:docker-nulldb#diff-3254677a7917c6c01f55212f86c57fbfR62

File: bundler.d/nulldb.rb (hunk 1 above)
Comment (ehelms) [reply bundler.d/nulldb.rb:1]: Nah, comes down to packaging. In the RPM environment the PR includes this file only in `foreman-plugin` sub-package which is used for builds only.

**Review by tbrisker — APPROVED** [review]
Fine by me, let's get this train rolling. Thanks @ehelms !

**Review by mmoll — APPROVED** [review]
thanks @ehelms for tackling this!

File: bundler.d/nulldb.rb (hunk 1 above)
Comment (ekohl) [reply bundler.d/nulldb.rb:1]: I think that if `foreman-build` includes the file, it should be ok. Then it can be a group.

**Review by ohadlevy — APPROVED** [review]

**Comment by tbrisker** [comment]
Thanks @ehelms , packaging has been merged

**Pull request merged**